src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+/// the quantized value remains within the error bound of the original value. If the
+/// `error_bound` is equal to zero, the value is directly returned.
+fn quantize(value: f64, error_bound: f32) f64 {
+    if (error_bound != 0)
+        return @floor(value / error_bound) * error_bound;
+    return value;
+}
+
+/// Appends the `metadata` to the hash map `metadata_map`. The `allocator` to use for allocating
+/// the memory for a new ArrayList if the `metadata.interception` does not exist.
+fn appendSegmentMetadata(
+    metadata_map: *HashMap(
+        f64,
+        ArrayList(SegmentMetadata),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ),
+    metadata: SegmentMetadata,
+    allocator: mem.Allocator,
+) !void {
+    const get_result = try metadata_map.getOrPut(metadata.interception);
+    if (!get_result.found_existing) {
+        get_result.value_ptr.* = ArrayList(SegmentMetadata).init(allocator);
+    }
+    try get_result.value_ptr.*.append(metadata);
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `lower_bound_slope`.
+fn compareMetadataBySlope() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
+            return metadata_one.lower_bound_slope < metadata_two.lower_bound_slope;
+        }
+    }.inner;
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `start_time`.
+fn compareMetadataByStartTime() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
+            return metadata_one.start_time < metadata_two.start_time;
+        }
+    }.inner;
+}
+
+/// Append `value` of `type` determined at compile time to `compressed_values`.
+fn appendValue(comptime T: type, value: T, compressed_values: *std.ArrayList(u8)) !void {
+    // Compile-time type check
+    switch (@TypeOf(value)) {
+        f64, usize => {
+            const value_as_bytes: [8]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        f32 => {
+            const value_as_bytes: [4]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        else => @compileError("Unsupported type for append value function"),
+    }
+}
+
+test "f64 context can hash" {
+    const allocator = testing.allocator;
+    var f64_hash_map = HashMap(
+        f64,
+        f64,
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ).init(allocator);
+    defer f64_hash_map.deinit();
+    var rnd = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
+
+    // Even a small deviation like this should not impact the ability to hash the numbers.
+    const deviation_margin = 1e-16;
```

Comment:
He purposely of the margin is unclear, maybe extend the comment.

Reply:
Fixed.